qiskit/pulse/ir/ir.py

```diff
@@ -0,0 +1,375 @@
+            initial_time (Optional): Starting time of the instruction. Defaults to ``None``
+
+        Raises:
+            PulseError: if ``initial_time`` is not ``None`` and not non-negative integer.
+        """
+        if initial_time is not None and (
+            not isinstance(initial_time, (int, np.integer)) or initial_time < 0
+        ):
+            raise PulseError("initial_time must be a non-negative integer.")
+
+        self._instruction = instruction
+        self._initial_time = initial_time
+
+    @property
+    def instruction(self) -> Instruction:
+        """Return the instruction associated with the IrInstruction"""
+        return self._instruction
+
+    @property
+    def initial_time(self) -> int | None:
+        return self._initial_time
+
+    @property
+    def duration(self) -> int:
+        return self._instruction.duration
+
+    @initial_time.setter
+    def initial_time(self, value: int):
+        """Set ``initial_time``
+
+        Args:
+            value: The integer value of ``initial_time``.
+
+        Raises:
+            PulseError: if ``value`` is not ``None`` and not non-negative integer.
+        """
+        if not isinstance(value, (int, np.integer)) or value < 0:
+            raise PulseError("initial_time must be a non-negative integer")
+        self._initial_time = value
+
+    def shift_initial_time(self, value: int):
+        """Shift ``initial_time``
+
+        Shifts ``initial_time`` to ``initial_time+value``.
+
+        Args:
+            value: The integer value by which ``initial_time`` is to be shifted.
+
+        Raises:
+            PulseError: If the instruction is not scheduled.
+        """
+        if self.initial_time is None:
+            raise PulseError("Can not shift initial_time of an untimed element")
+
+        # validation of new initial_time is done in initial_time setter.
+        self.initial_time = self.initial_time + value
+
+    def __eq__(self, other: "IrInstruction") -> bool:
+        """Return True iff self and other are equal
+
+        Args:
+            other: The IR instruction to compare to this one.
+
+        Returns:
+            True iff equal.
+        """
+        return (
+            type(self) is type(other)
+            and self._instruction == other.instruction
+            and self._initial_time == other.initial_time
+        )
+
+    def __repr__(self) -> str:
+        """IrInstruction representation"""
+        return f"IrInstruction({self.instruction}, {self.initial_time})"
+
+
+class IrBlock(IrElement):
+    """
+    ``IrBlock`` is the backbone of the  intermediate representation used in the Qiskit Pulse compiler.
+    A pulse program is represented as a single ``IrBlock`` object, with elements
+    which include IR instructions and other
+    nested ``IrBlock`` objects. This structure mimics that of :class:`.qiskit.pulse.ScheduleBlock`
+    which is the main pulse program format used in Qiskit.
+    """
```

Comment:
```suggestion
    """Pulse IR for sequence.
    
    ``IrBlock`` is the backbone of the  intermediate representation used in the Qiskit Pulse compiler.
    A pulse program is represented as a single ``IrBlock`` object, with elements
    which include IR instructions and other nested ``IrBlock`` objects.
    """
```
You don't need to tie the IR with existing frontend.